osu.Game.Modes.Taiko/Objects/Drawables/DrawableHitCircle.cs

```diff
@@ -0,0 +1,196 @@
+
+        private CirclePiece bodyPiece;
+        private Container bodyContainer;
+
+        private bool validKeyPressed = true;
+
+        public DrawableHitCircle(HitCircle hitCircle)
+            : base(hitCircle)
+        {
+            Size = new Vector2(TaikoHitObject.CIRCLE_RADIUS * 2);
+
+            Children = new Drawable[]
+            {
+                bodyContainer = new Container
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    
+                    Children = new[]
+                    {
+                        bodyPiece = CreateBody()
+                    }
+                }
+            };
+
+            bodyPiece.Kiai = hitCircle.Kiai;
+        }
+
+        /// <summary>
+        /// Creates a body circle of this HitCircle.
+        /// </summary>
+        /// <returns>The body circle.</returns>
+        protected abstract CirclePiece CreateBody();
+
+        protected override void CheckJudgement(bool userTriggered)
+        {
+            if (!userTriggered)
+            {
+                if (Judgement.TimeOffset > HitObject.HitWindowGood)
+                    Judgement.Result = HitResult.Miss;
+                return;
+            }
+
+            double hitOffset = Math.Abs(Judgement.TimeOffset);
+
+            // Must be within great range to be hittable/missable
+            if (hitOffset > HitObject.HitWindowMiss)
+                return;
+
+            TaikoJudgementInfo tji = Judgement as TaikoJudgementInfo;
+
+            if (!validKeyPressed)
+                Judgement.Result = HitResult.Miss;
+            else if (hitOffset < HitObject.HitWindowGood)
+            {
+                Judgement.Result = HitResult.Hit;
+
+                if (hitOffset < HitObject.HitWindowGreat)
+                    tji.Score = TaikoScoreResult.Great;
+                else
+                    tji.Score = TaikoScoreResult.Good;
+            }
+            else
+                Judgement.Result = HitResult.Miss;
+        }
+
+        protected override void UpdateState(ArmedState state)
+        {
+            if (!IsLoaded)
+                return;
+
+            base.UpdateState(state);
+
+            switch (State)
+            {
+                case ArmedState.Idle:
+                    break;
+                case ArmedState.Miss:
+                    bodyContainer.FadeColour(Color4.Red, 100, EasingTypes.OutQuint);
+                    bodyContainer.FadeOut(100);
+                    break;
+                case ArmedState.Hit:
+                    bodyContainer.ScaleTo(1.5f, 150, EasingTypes.OutQuint);
+                    bodyContainer.FadeOut(150);
+                    break;
+            }
+        }
+
+        protected override void Update()
+        {
+            // If the HitCircle was hit, make it stop moving
+            if (State != ArmedState.Hit)

```

Comment:
This is dangerous, as it blocks *everything* in the underlying Update. Consider making a property `ShouldMove` or just moving this conditional to the base class. Also I'd move it out to an explicit updateMovement() method.